scipy/sparse/_coo.py

```diff
@@ -109,112 +128,130 @@ def reshape(self, *args, **kwargs):
             else:
                 return self
 
-        nrows, ncols = self.shape
-
-        if order == 'C':
-            # Upcast to avoid overflows: the coo_array constructor
-            # below will downcast the results to a smaller dtype, if
-            # possible.
-            maxval = (ncols * max(0, nrows - 1) + max(0, ncols - 1))
-            dtype = self._get_index_dtype(maxval=maxval)
-
-            flat_indices = np.multiply(ncols, self.row, dtype=dtype) + self.col
-            new_row, new_col = divmod(flat_indices, shape[1])
-        elif order == 'F':
-            maxval = (nrows * max(0, ncols - 1) + max(0, nrows - 1))
-            dtype = self._get_index_dtype(maxval=maxval)
-
-            flat_indices = np.multiply(nrows, self.col, dtype=dtype) + self.row
-            new_col, new_row = divmod(flat_indices, shape[0])
+        # When reducing the number of dimensions, we need to be careful about
+        # index overflow. This is why we can't simply call
+        # `np.ravel_multi_index()` followed by `np.unravel_index()` here.
+        flat_indices = _ravel_indices(self.indices, self.shape, order=order)
```

Comment:
Is this a bug/lack of functionality in `ravel_multi_index`? If so, happy to file a NumPy issue.